ni_measurementlink_service/_internal/discovery_client.py

```diff
@@ -204,12 +206,63 @@ def resolve_service(self, provided_interface: str, service_class: str = "") -> S
 
 def _get_discovery_service_address() -> str:
     key_file_path = _get_key_file_path()
+    _ensure_discovery_service_started(key_file_path)
     _logger.debug("Discovery service key file path: %s", key_file_path)
     with _open_key_file(str(key_file_path)) as key_file:
         key_json = json.load(key_file)
         return "localhost:" + key_json["InsecurePort"]
 
 
+def _ensure_discovery_service_started(key_file_path: pathlib.Path):
+    """Make sure discovery service is started.
+
+    Check whether discovery service key file exists, if not start the discovery
+    service.
+
+    """
+    if _key_file_exists(key_file_path):
+        return
+
+    exe_file_path = _get_discovery_service_location()
+    _start_service(exe_file_path)
+
+
+def _get_discovery_service_location() -> pathlib.PurePath:
+    """Gets the location of the discovery service process executable.
+
+    Returns:
+    -------
+        exe_file_path: path to the discovery service exe which can be used
+                to activate the discovery service.
+
+    """
+    registration_json_path = _get_registration_json_file_path()
+    registration_json_obj = json.load(open(str(registration_json_path)))
+    root_directory = pathlib.PurePath(registration_json_path).parent
+    exe_file_path = root_directory.joinpath(registration_json_obj["discovery"]["path"])
+    return exe_file_path
+
+
+def _get_registration_json_file_path():
+    return (
+        pathlib.Path(os.environ["ProgramFiles"])
+        / "National Instruments"
+        / "Shared"
+        / "MeasurementLink"
+        / "MeasurementLinkServices.json"
+    )
+
+
+def _key_file_exists(key_file_path) -> bool:
+    return pathlib.Path(key_file_path).is_file()
+
+
+def _start_service(exe_file_path):
+    """Starts the service at the specified path and wait for the service to get up and running."""
+    subprocess.Popen([exe_file_path], cwd=exe_file_path.parent)
+    time.sleep(1)
```

Comment:
should we sleep for 1 second or wait until the discoveryService.json is created?

Reply:
See 'WaitForKeyFileExists' in ASW.
https://ngsourcebrowser.ni.systems:4111/#NationalInstruments.MeasurementLink.Core/DiscoveryClient.cs,096b739b73ca2b5a,references

Reply:
Also note that in the C# version, the 30-second timeout is handled outside the `WaitForKeyFileExists` method using a `CancellationToken`. In Python, you will have to handle this in the polling loop.

Reply:
Updated the code to wait for key file with timeout of 30 seconds

Reply:
Replaced sleep method with a polling loop which checks for key file existence every 0.1 second and 30-second timeout of the check is also handled.